### ISSUE TITLE
sass-lint is running into fatal errors with a fresh container rebuild

### ISSUE THREAD

**Issue by lmorchard** [mozilla/testpilot#633]
TL;DR: Something changed in `sass-lint` that doesn't like our `.scss` files now. It's breaking my local dev machine.

Because `gulp-sass` and `gulp-sass-lint` both have fuzzy version requirements for their dependencies, we end up getting incompatible upgrades to their dependencies that don't like our code.

This problem won't be seen on Circle CI until/unless we clear the dependency caches, but you'll see it if you rebuild all your docker containers. You should also see if it you install gulp from scratch locally on the host (e.g. `rm -rf node_modules && npm install`)

More details:

I just rebuilt all my docker containers totally from scratch. I removed all the containers and even deleted & recreated my docker-machine on OS X. Now, I'm getting errors like this in gulp:

```
frontend_watcher_1 | [21:01:07] Starting 'sass-lint'...
frontend_watcher_1 |
frontend_watcher_1 | testpilot/frontend/static-src/styles/_layout.scss
frontend_watcher_1 |   60:1  error  Please check validity of the block starting from line #60  Fatal
frontend_watcher_1 |
frontend_watcher_1 | ✖ 1 problem (1 error, 0 warnings)
frontend_watcher_1 |
frontend_watcher_1 |
frontend_watcher_1 | testpilot/frontend/static-src/styles/_state.scss
frontend_watcher_1 |   128:1  error  Please check validity of the block starting from line #128  Fatal
frontend_watcher_1 |
frontend_watcher_1 | ✖ 1 problem (1 error, 0 warnings)
```

This doesn't happen on Circle CI, so I started a build with ssh and looked at what was different. Here's what I think happened:

`gulp-sass-lint` depends on `sass-lint`. And, until recently, `gulp-sass-lint` was pulling in `sass-lint` v1.5.1. But, the `package.json` in `gulp-sass-lint` specifies `"sass-lint": "^1.3.2"`. So, since `sass-lint` v1.6.0 is now available and the '^' version spec means an upgrade is okay, so fresh installations use the newest version.

Unfortunately, it looks like something changed between v1.5.1 and v1.6.0 that doesn't agree with our `.scss` files.

**Comment by lmorchard** [comment]
Looks like we might be hitting this bug in the upstream dependency:

https://github.com/sasstools/sass-lint/issues/642

**Comment by lmorchard** [comment]
Also looks like when I forcibly downgrade to sass-lint 1.5.1, it doesn't like [this ampersand in _layout.scss](https://github.com/mozilla/testpilot/blame/master/testpilot/frontend/static-src/styles/_layout.scss#L60):

```
frontend_watcher_1 | [21:14:01] Starting 'styles'...
frontend_watcher_1 | Error in plugin 'sass'
frontend_watcher_1 | Message:
frontend_watcher_1 |     testpilot/frontend/static-src/styles/_layout.scss
frontend_watcher_1 | Error: Invalid CSS after ":not(#how-to)": expected "{", was "&"
frontend_watcher_1 |
frontend_watcher_1 |        "&" may only be used at the beginning of a compound selector.
frontend_watcher_1 |         on line 60 of testpilot/frontend/static-src/styles/_layout.scss
frontend_watcher_1 | >>   .blue :not(#how-to)& {
frontend_watcher_1 |    ---------------------^
frontend_watcher_1 |
frontend_watcher_1 | [21:14:01] Finished 'styles' after 211 ms
```

I don't grok scss syntax well enough to know whether that _should_ be valid, or whether it's a typo. It got merged, and it passed Circle CI, so I assume it works for _some_ combination of sass code.

@johngruen / @meandavejustice - any thoughts on this stuff?

**Comment by johngruen** [comment]
@lmorchard yeah i think we can just rewrite this as `.blue &:not(#how-to)`

**Comment by johngruen** [comment]
PS. Thanks Obama

**Comment by lmorchard** [comment]
Poking at this second error, it looks like Circle CI is using a cached `gulp-sass` 2.2.0 with a `node-sass` dependency at v3.4.2 which in turn uses `libsass` v3.3.2

But my fresh local dev machine is using `gulp-sass` 2.3.0 -> `node-sass` v3.5.3 -> `libsass` v3.3.5.

Sometimes, I hate computers.

Edit: Yup, if I shell into my frontend_watcher docker container and forcibly downgrade gulp-sass' node-sass down to 3.4.2 by hand, this second error goes away.

**Comment by pdehaan** [comment]
https://medium.com/@xzyfer/why-node-sass-broke-your-code-and-semver-1b3e409c57b9#.zf1lkazie

**Comment by meandavejustice** [comment]
Thanks @lmorchard for tracking this down. And @pdehaan for providing context.

**Comment by johngruen** [comment]
@lmorchard we could also just convert everything to LESS if that's easier

**Comment by huajinwu** [comment]
IE6&IE7 CSS hack error! 
For example: *display: inline; 
Error message:
severity: '错误'
message: 'Please check validity of the block starting from line #21 (Fatal)'
at: '21,1'
source: 'sass-lint'
code: 'undefined'
